File: lab_1/src/InputParser.cpp
#include "InputParser.hpp"
#include <fstream>
#include <iostream>
#include <iterator>

InputParser::InputParser(int argc, char* argv[]) : isValid_(false) {
    if (argc != 4) {
        std::cerr << "Usage: <FA|KMP> <pattern> <filename>\n";
        return;
    }
    algorithm_ = argv[1];
    pattern_ = argv[2];
    filename_ = argv[3];

    // Wczytaj plik do tekstu
    std::ifstream file(filename_);
    if (!file) {
        std::cerr << "Cannot open file: " << filename_ << "\n";
        return;
    }
    text_ = std::string((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());

    // Zbuduj alfabet na podstawie tekstu i wzorca
    for (char c : text_) alphabet_.insert(c);
    for (char c : pattern_) alphabet_.insert(c);

    // Build UTF-8-aware alphabet: split text_ and pattern_ into UTF-8 codepoints
    auto extractUTF8 = [](const std::string &s) {
        std::set<std::string> out;
        for (size_t i = 0; i < s.size(); ) {
            unsigned char c = static_cast<unsigned char>(s[i]);
            size_t len = 1;
            if ((c & 0x80) == 0x00) len = 1;
            else if ((c & 0xE0) == 0xC0) len = 2;
            else if ((c & 0xF0) == 0xE0) len = 3;
            else if ((c & 0xF8) == 0xF0) len = 4;
            else len = 1; // fallback
            if (i + len <= s.size()) {
                out.insert(s.substr(i, len));
            }
            i += len;
        }
        return out;
    };
    // merge into a member container by reusing alphabet_? We'll store UTF-8 set in filename_ not allowed; instead keep a private static store via a mutable static? Simpler: keep a private member not declared - add one.

    isValid_ = true;
}

bool InputParser::validate() const { return isValid_; }
std::string InputParser::getAlgorithm() const { return algorithm_; }
std::string InputParser::getPattern() const { return pattern_; }
std::string InputParser::getText() const { return text_; }
std::set<char> InputParser::getAlphabet() const { return alphabet_; }
std::string InputParser::getFilename() const { return filename_; }

std::set<std::string> InputParser::getAlphabetUTF8() const {
    // Build sets on the fly from stored text_ and pattern_
    auto extract = [](const std::string &s) {
        std::set<std::string> out;
        for (size_t i = 0; i < s.size(); ) {
            unsigned char c = static_cast<unsigned char>(s[i]);
            size_t len = 1;
            if ((c & 0x80) == 0x00) len = 1;
            else if ((c & 0xE0) == 0xC0) len = 2;
            else if ((c & 0xF0) == 0xE0) len = 3;
            else if ((c & 0xF8) == 0xF0) len = 4;
            else len = 1;
            if (i + len <= s.size()) out.insert(s.substr(i, len));
            i += len;
        }
        return out;
    };

    std::set<std::string> res = extract(text_);
    auto pat = extract(pattern_);
    res.insert(pat.begin(), pat.end());
    return res;
}

File: lab_1_python/src/input_parser.py
import sys
from typing import Set


class InputParser:
	"""Minimal positional parser mirroring the C++ InputParser.

	Usage (positional):
	  python main.py <FA|KMP> <pattern> <filename>

	The constructor expects argv (full argv including script name) or uses sys.argv.
	"""

	def __init__(self, argv=None):
		argv = argv if argv is not None else sys.argv
		self.is_valid = False
		self.algorithm = ""
		self.pattern = ""
		self.text = ""
		self.alphabet = set()
		self.filename = ""

		# Expect exactly: [script, algorithm, pattern, filename]
		if len(argv) != 4:
			print("Usage: <FA|KMP> <pattern> <filename>", file=sys.stderr)
			return

		self.algorithm = argv[1]
		self.pattern = argv[2]
		self.filename = argv[3]

		# Read file to text
		try:
			with open(self.filename, 'r', encoding='utf-8', errors='replace') as f:
				self.text = f.read()
		except Exception as e:
			print(f"Cannot open file: {self.filename} ({e})", file=sys.stderr)
			return

		# Normalize CRLF (remove CR)
		self.pattern = self.pattern.replace('\r', '')
		self.text = self.text.replace('\r', '')

		# Build alphabet from text and pattern
		for c in self.text:
			self.alphabet.add(c)
		for c in self.pattern:
			self.alphabet.add(c)

		self.is_valid = True

	def validate(self) -> bool:
		return self.is_valid

	def get_algorithm(self) -> str:
		return self.algorithm

	def get_pattern(self) -> str:
		return self.pattern

	def get_text(self) -> str:
		return self.text

	def get_alphabet(self) -> Set[str]:
		return set(self.alphabet)

	def get_filename(self) -> str:
		return self.filename

